modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java

```diff
@@ -182,4 +212,160 @@ private BinaryMetadata readMetadata(int typeId) {
+        /** */
+        private final BlockingQueue<WriteOpTask> queue = new LinkedBlockingQueue();
+
+        /** */
+        BinaryMetadataAsyncWriter() {
+            super(ctx.igniteInstanceName(), "binary-metadata-writer", BinaryMetadataFileStore.this.log, ctx.workersRegistry());
+        }
+
+        /** */
+        void submit(WriteOpTask task) {
+            if (isCancelled())
+                return;
+
+            writeOpFutures.put(new WriteOpSync(task.meta.typeId(), task.typeVer), new GridFutureAdapter());
+
+            queue.add(task);
+        }
+
+        /** {@inheritDoc} */
+        @Override public void cancel() {
+            super.cancel();
+
+            queue.clear();
+
+            IgniteCheckedException err = new IgniteCheckedException("Operation has been cancelled (node is stopping).");
+
+            for (GridFutureAdapter fut : writeOpFutures.values())
+                fut.onDone(err);
+
+            writeOpFutures.clear();
+        }
+
+        /** {@inheritDoc} */
+        @Override protected void body() throws InterruptedException, IgniteInterruptedCheckedException {
+           while (!isCancelled()) {
+               try {
+                   body0();
+               }
+               catch (InterruptedException e) {
+                   if (!isCancelled) {
+                       ctx.failure().process(new FailureContext(FailureType.SYSTEM_WORKER_TERMINATION, e));
+
+                       throw e;
+                   }
+               }
+           }
+        }
+
+        /** */
+        private void body0() throws InterruptedException {
+            WriteOpTask task;
+
+            blockingSectionBegin();
+
+            try {
+                task = queue.take();
+
+                writeMetadata(task.meta);
+            }
+            finally {
+                blockingSectionEnd();
+            }
+
+            GridFutureAdapter fut = writeOpFutures.remove(new WriteOpSync(task.meta.typeId(), task.typeVer));
+
+            if (fut != null)
+                fut.onDone();
+        }
+    }
+
+    /**
+     *
+     */
+    private static final class WriteOpTask {
+        /** */
+        private final BinaryMetadata meta;
+        /** */
+        private final int typeVer;
+
+        /** */
+        private WriteOpTask(BinaryMetadata meta, int ver) {
+            this.meta = meta;
+            typeVer = ver;
+        }
+    }
+
+    /**
+     *
+     */
+    private static final class WriteOpSync {
+        /** */
+        private final int typeId;
+
+        /** */
+        private final int typeVer;
+
+        /** */
+        private WriteOpSync(int typeId, int typeVer) {
+            this.typeId = typeId;
+            this.typeVer = typeVer;
+        }
+
+        /** {@inheritDoc} */
+        @Override public int hashCode() {
+            return Objects.hash(typeId, typeVer);
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean equals(Object obj) {
+            if (obj == this)
```

Comment:
Looking at your code I guarantee that this condition is always false.

Reply:
Removed.